pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,371 @@
+
+	cmUpdate := unstructuredFor(configMapGVK, cm1Name)
+	cmUpdate.SetLabels(map[string]string{"testlabel": "test"}) // trigger an instance update
+	require.NoError(t, c.Update(ctx, cmUpdate))
+
+	require.Eventually(t, func() bool {
+		instance := cfClient.GetData(cmKey)
+		unInstance, ok := instance.(*unstructured.Unstructured)
+		require.True(t, ok)
+
+		value, found, err := unstructured.NestedString(unInstance.Object, "metadata", "labels", "testlabel")
+		require.NoError(t, err)
+
+		return found && "test" == value
+	}, eventuallyTimeout, eventuallyTicker)
+}
+
+func deleteResource(ctx context.Context, c client.Client, resounce *unstructured.Unstructured) error {
+	err := c.Delete(ctx, resounce)
+	if apierrors.IsNotFound(err) {
+		// resource does not exist, this is good
+		return nil
+	}
+
+	return err
+}
+
+func expectedCheck(cfClient *cachemanager.FakeCfClient, expected map[cachemanager.CfDataKey]interface{}) func() bool {
+	return func() bool {
+		if cfClient.Len() != len(expected) {
+			return false
+		}
+		if cfClient.Contains(expected) {
+			return true
+		}
+
+		return false
+	}
+}
+
+func unstructuredFor(gvk schema.GroupVersionKind, name string) *unstructured.Unstructured {
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(gvk)
+	u.SetName(name)
+	u.SetNamespace("default")
+	if gvk.Kind == "Pod" {
+		u.Object["spec"] = map[string]interface{}{
+			"containers": []map[string]interface{}{
+				{
+					"name":  "foo-container",
+					"image": "foo-image",
+				},
+			},
+		}
+	}
+	return u
+}
+
+type testResources struct {
+	*cachemanager.CacheManager
+	cachemanager.CFDataClient
+	*aggregator.GVKAgreggator
+}
+
+func makeTestResources(t *testing.T, mgr manager.Manager, wm *watch.Manager, reader client.Reader) (testResources, context.Context) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	cfClient := &cachemanager.FakeCfClient{}
+	tracker, err := readiness.SetupTracker(mgr, false, false, false)
+	require.NoError(t, err)
+	processExcluder := process.Get()
+	processExcluder.Add([]configv1alpha1.MatchEntry{
+		{
+			ExcludedNamespaces: []wildcard.Wildcard{"kube-system"},
+			Processes:          []string{"sync"},
+		},
+	})
+	events := make(chan event.GenericEvent, 1024)
+	w, err := wm.NewRegistrar(
+		"test-cache-manager",
+		events)
+	require.NoError(t, err)
+
+	aggregator := aggregator.NewGVKAggregator()
+	config := &cachemanager.Config{
+		CfClient:         cfClient,
+		SyncMetricsCache: syncutil.NewMetricsCache(),
+		Tracker:          tracker,
+		ProcessExcluder:  processExcluder,
+		WatchedSet:       watch.NewSet(),
+		Registrar:        w,
+		Reader:           reader,
+		GVKAggregator:    aggregator,
+	}
+	cacheManager, err := cachemanager.NewCacheManager(config)
+	require.NoError(t, err)
+
+	syncAdder := syncc.Adder{
+		Events:       events,
+		CacheManager: cacheManager,
+	}
+	require.NoError(t, syncAdder.Add(mgr), "registering sync controller")
+	go func() {
+		require.NoError(t, cacheManager.Start(ctx))
```

Comment:
require cannot be used in child goroutines: https://github.com/stretchr/testify/issues/772